### DISCUSSION TITLE
Interfacing Gurobi?

### DISCUSSION THREAD

**Discussion by gy2256** [casadi/casadi#3398] [Q&A] [answered]
How do I use Gurobi with OptiStack? I'm trying to understand the adaptor. I found there's a `libgurobi_adaptor.dylib `in Casadi, but I don't know how to use it. There's no documentation on how to interface Casadi with third-party solvers. Any help would be appreciated. 

This is the error message that I have received when I try to use Gurobi:
```
CasADi - 2023-10-04 07:35:07 WARNING("Failed to load Gurobi adaptor: Gurobi load adaptor needs an environmental variable <GUROBI_VERSION> such that libgurobi<GUROBI_VERSION>.dylib can be found..") [.../casadi/interfaces/gurobi/gurobi_interface.cpp:45]
CasADi - 2023-10-04 07:35:07 WARNING(".../casadi/core/plugin_interface.hpp:170: Assertion "flag==0" failed:
Registration of plugin failed.") [.../casadi/core/plugin_interface.hpp:142]
Traceback (most recent call last):
  File "/Users/mac/test.py", line 9, in <module>
    opti.solve()
  File "/opt/anaconda3/envs/rl-mpc/lib/python3.10/site-packages/casadi/casadi.py", line 47783, in solve
    return _casadi.Opti_solve(self, *args)
RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157:
.../casadi/core/plugin_interface.hpp:292: Plugin 'gurobi' is not found.
```

**Comment by GianlucaGiacomelli** [comment]
Me too I am struggling a lot on an interface between Casadi OptiStack and Gurobi

**Reply by gy2256** [reply]
Yeah. There must be away to interface it. Here is the outdated instruction: [https://github.com/casadi/casadi/wiki/FAQ:-how-to-get-third-party-solvers-to-work%3F](https://github.com/casadi/casadi/wiki/FAQ:-how-to-get-third-party-solvers-to-work%3F). But I could not get that one working either.

**Comment by jgillis** [comment]
I'll make an overdue update to the wiki. What error messages are you getting when doing this:

```python
opti=Opti('conic')
x=opti.variable()
opti.minimize((x-1)*x)
opti.subject_to(x>=3)
opti.solver('gurobi')
opti.solve()
```

**Reply by gy2256** [reply]
Thanks for the help. Just a quick background of my setup: I am using a MacBook Pro with Apple M1 Pro SOC and Mac OS Sonoma. 

I have installed the latest Gurobi and I confirm that `libgurobi100.dylib` is in `/usr/local/lib`. 


I received the following message:
```
CasADi - 2023-10-04 07:35:07 WARNING("Failed to load Gurobi adaptor: Gurobi load adaptor needs an environmental variable <GUROBI_VERSION> such that libgurobi<GUROBI_VERSION>.dylib can be found..") [.../casadi/interfaces/gurobi/gurobi_interface.cpp:45]
CasADi - 2023-10-04 07:35:07 WARNING(".../casadi/core/plugin_interface.hpp:170: Assertion "flag==0" failed:
Registration of plugin failed.") [.../casadi/core/plugin_interface.hpp:142]
Traceback (most recent call last):
  File "/Users/mac/test.py", line 9, in <module>
    opti.solve()
  File "/opt/anaconda3/envs/rl-mpc/lib/python3.10/site-packages/casadi/casadi.py", line 47783, in solve
    return _casadi.Opti_solve(self, *args)
RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157:
.../casadi/core/plugin_interface.hpp:292: Plugin 'gurobi' is not found.
```

**Reply by jgillis** [reply, marked as answer]
Have you tried setting GUROBI_VERSION env var to 100?

**Reply by gy2256** [reply]
It worked after I placed `export GUROBI_VERSION="100"` in my `~/.zshrc`! Thanks for the help. 

For the record, here is the output after running your code from above:
```
Academic license - for non-commercial use only - expires 2024-05-16
Gurobi Optimizer version 10.0.2 build v10.0.2rc0 (mac64[arm])

CPU model: Apple M1 Pro
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 1 rows, 1 columns and 1 nonzeros
Model fingerprint: 0xeceb4715
Model has 1 quadratic objective term
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [1e+00, 1e+00]
  QObjective range [2e+00, 2e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+00, 3e+00]
Presolve removed 1 rows and 1 columns
Presolve time: 0.00s
Presolve: All rows and columns removed

Barrier solved model in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective 6.00000000e+00
```

**Reply by jgillis** [reply]
Do suggest an alternative description for that warning that would have helped you earlier

**Reply by rtapia8484** [reply]
I have the same issue, except that I can get it to work with MATLAB but not when using python:
CasADi - 2024-09-13 16:55:43 WARNING("Failed to load Gurobi adaptor: Could not find library 'gurobi100.dll' (gurobi<GUROBI_VERSION>.dll). Consider adding the appropriate gurobi folder to environmental variable 'PATH', or specifying a different variable 'CPLEX_VERSION'..") [.../casadi/interfaces/gurobi/gurobi_interface.cpp:45]
CasADi - 2024-09-13 16:55:43 WARNING(".../casadi/core/plugin_interface.hpp:170: Assertion "flag==0" failed:
Registration of plugin failed.") [.../casadi/core/plugin_interface.hpp:142]


I have the GUROBI_VERSION variable defined, as well as GUROBI_HOME

**Reply by rtapia8484** [reply]
Just need to add the following to python:
import gurobipy as gp

**Reply by polcrub** [reply]
> It worked after I placed `export GUROBI_VERSION="100"` in my `~/.zshrc`! Thanks for the help.
> 
> For the record, here is the output after running your code from above:
> 
> ```
> Academic license - for non-commercial use only - expires 2024-05-16
> Gurobi Optimizer version 10.0.2 build v10.0.2rc0 (mac64[arm])
> 
> CPU model: Apple M1 Pro
> Thread count: 8 physical cores, 8 logical processors, using up to 8 threads
> 
> Optimize a model with 1 rows, 1 columns and 1 nonzeros
> Model fingerprint: 0xeceb4715
> Model has 1 quadratic objective term
> Coefficient statistics:
>   Matrix range     [1e+00, 1e+00]
>   Objective range  [1e+00, 1e+00]
>   QObjective range [2e+00, 2e+00]
>   Bounds range     [0e+00, 0e+00]
>   RHS range        [3e+00, 3e+00]
> Presolve removed 1 rows and 1 columns
> Presolve time: 0.00s
> Presolve: All rows and columns removed
> 
> Barrier solved model in 0 iterations and 0.00 seconds (0.00 work units)
> Optimal objective 6.00000000e+00
> ```

Hello, how did you achieve to get the GUROBI solution displayed?? I can't achieve such logging with the opts.verbose variable.. Moreover, is there any way to access the GUROBI-specific outputs? as can be the MIPGap achieved.

**Comment by gy2256** [comment]
I have another question. Does the Casadi support solving nonlinear optimization with Gurobi?  I cannot solve the following problem 

```
from casadi import Opti

opti=Opti('conic')

x = opti.variable()
y = opti.variable()

opti.minimize(  (y-x**2)**2   )
opti.subject_to( x**2+y**2==1 )
opti.subject_to(       x+y>=1 )

opti.solver('gurobi')
opti.solve()
```

**Reply by aghezz1** [reply]
Just overlooked this question. No, casadi interface to gurobi is only via conic or qpsol.
So atm yo can't solve NLP via gurobi.

**Comment by jonarriza96** [comment]
For casadi 3.6.5 and gurobi 12.0 --> `export GUROBI_VERSION="120"`
